tests/core/vm/test_PAYGAS_tx_fee.py

```diff
@@ -34,16 +35,162 @@ def test_trigger_PAYGAS(shard_chain_without_block_validation):  # noqa: F811
+    gas_price = 33
     vrs = 64 * b'\xAA' + b'\x01'
-    trigger_tx = new_sharding_transaction(tx_initiator, recipient, amount, gas_price, vrs, b'')
+    trigger_tx = new_sharding_transaction(
+        tx_initiator,
+        recipient,
+        amount,
+        bytes([gas_price]),
+        vrs,
+        b''
+    )
 
+    with vm.state.state_db(read_only=True) as state_db:
+        balance_before_trigger = state_db.get_balance(simple_forwarder_contract['address'])
     computation, _ = vm.apply_transaction(trigger_tx)
     assert not computation.is_error
     gas_used = vm.block.header.gas_used - last_gas_used
     assert gas_used > trigger_tx.intrinsic_gas
     last_gas_used = vm.block.header.gas_used
     with vm.state.state_db(read_only=True) as state_db:
+        assert state_db.get_balance(tx_initiator) == balance_before_trigger - gas_used * gas_price
         assert state_db.get_balance(recipient) == amount
+
+
+def test_PAYGAS_edge_cases(shard_chain_without_block_validation):  # noqa: F811
+    # Case 1: PAYGAS not triggered
+    chain = shard_chain_without_block_validation
+    vm = chain.get_vm()
+
+    forwarder_addr = simple_forwarder_contract['address']
+    forwarder_contract_deploy_tx = new_sharding_transaction(
+        tx_initiator=forwarder_addr,
+        data_destination=b'',
+        data_value=0,
+        data_msgdata=b'',
+        data_vrs=b'',
+        code=simple_forwarder_contract['bytecode'],
+    )
+    computation, _ = vm.apply_transaction(forwarder_contract_deploy_tx)
+    assert not computation.is_error
+
+    # Trigger the forwarder contract which does not have PAYGAS opcode
+    recipient = decode_hex('0xa94f5374fce5edbc8e2a8697c15331677e6ebf0c')
+    amount = 0
+    tx_initiator = forwarder_addr
+    gas_price = bytes([0])
+    vrs = 64 * b'\xAA' + b'\x01'
+    trigger_tx = new_sharding_transaction(tx_initiator, recipient, amount, gas_price, vrs, b'')
+
+    with vm.state.state_db(read_only=True) as state_db:
+        balance_before_trigger = state_db.get_balance(forwarder_addr)
+    computation, _ = vm.apply_transaction(trigger_tx)
+    assert not computation.is_error
+    with vm.state.state_db(read_only=True) as state_db:
+        # Check that balance of the contract is the same
+        assert balance_before_trigger == state_db.get_balance(forwarder_addr)
+    
+    # Case 2: PAYGAS triggered with 0 gas price
+    PAYGAS_contract_addr = PAYGAS_contract_normal['address']
+    normal_PAYGAS_contract_deploy_tx = new_sharding_transaction(
+        tx_initiator=PAYGAS_contract_addr,
+        data_destination=b'',
+        data_value=0,
+        data_msgdata=b'',
+        data_vrs=b'',
+        code=PAYGAS_contract_normal['bytecode'],
+    )
+    computation, _ = vm.apply_transaction(normal_PAYGAS_contract_deploy_tx)
+    assert not computation.is_error
+
+    # Trigger the PAYGAS contract with gas price set to 0
+    recipient = decode_hex('0xa94f5374fce5edbc8e2a8697c15331677e6ebf0c')
```

Comment:
I'd like to see these divided up into individual test cases rather than the current format (which I think is error prone and harder to follow).

Reply:
(or even using @jannikluhn 's fancy new test case format from #350 if you're feeling brave and @jannikluhn doesn't think it's a terrible idea.  Best way to validate an API is to use it in-my-experience.